src/frontend/metrics/DeviceDiagnosticMetrics.ts

```diff
@@ -0,0 +1,198 @@
+} from 'react-native';
+import {getRandomBytes} from 'expo-crypto';
+import {uint8ArrayToHex} from 'uint8array-extras';
+import * as NetInfo from '@react-native-community/netinfo';
+import * as Device from 'expo-device';
+import {getMonthlyHash} from './getMonthlyHash';
+import {sendMetricsData} from './sendMetricsData';
+import {getMetricsRequestInfo} from './getMetricsRequestInfo';
+import {storage} from '../hooks/persistedState/createPersistedState';
+import {
+  isDateValid,
+  isSameUtcMonthAndYear,
+  beginningOfNextMonthUtc,
+} from '../lib/date';
+import {setLongTimeout, clearLongTimeout} from '../lib/longTimeout';
+
+const STORAGE_KEY = 'DeviceDiagnosticMetricsLastSentAt';
+
+type DeviceDiagnosticMetricsData = {
+  type: 'device diagnostics v1';
+  monthlyDeviceHash: string;
+  brand?: string;
+  deviceType?: string;
+  isEmulator?: true;
+  manufacturer?: string;
+  model?: string;
+  os: string;
+  screen: {width: number; height: number; pixelRatio: number};
+  supportedCpuArchitectures?: string[];
+  totalMemory?: number;
+};
+
+function deviceTypeToString(deviceType: null | Device.DeviceType): string {
+  switch (deviceType) {
+    case Device.DeviceType.PHONE:
+      return 'PHONE';
+    case Device.DeviceType.TABLET:
+      return 'TABLET';
+    case Device.DeviceType.DESKTOP:
+      return 'DESKTOP';
+    case Device.DeviceType.TV:
+      return 'TV';
+    case null:
+    case Device.DeviceType.UNKNOWN:
+    default:
+      return 'UNKNOWN';
+  }
+}
+
+function getMetricsDeviceId(): string {
+  const result = storage.getString('MetricsDeviceId');
+  if (result) return result;
+
+  const newId = uint8ArrayToHex(getRandomBytes(16));
+  storage.set('MetricsDeviceId', newId);
+  return newId;
+}
+
+function setIfNotNull<ObjT, KeyT extends keyof ObjT>(
+  obj: ObjT,
+  key: KeyT,
+  value: null | ObjT[KeyT],
+): void {
+  if (value !== null) obj[key] = value;
+}
+
+async function generateDeviceDiagnosticMetricsData(): Promise<DeviceDiagnosticMetricsData> {
+  const screen = Dimensions.get('screen');
+
+  const result: DeviceDiagnosticMetricsData = {
+    type: 'device diagnostics v1',
+    os: Platform.OS,
+    screen: {
+      width: screen.width,
+      height: screen.height,
+      pixelRatio: PixelRatio.get(),
+    },
+    deviceType: deviceTypeToString(Device.deviceType),
+    monthlyDeviceHash: await getMonthlyHash(
+      'device diagnostics',
+      getMetricsDeviceId(),
+      new Date(),
+    ),
+  };
+
+  setIfNotNull(result, 'brand', Device.brand);
+  setIfNotNull(result, 'manufacturer', Device.manufacturer);
+  setIfNotNull(result, 'model', Device.modelId || Device.modelName);
+  setIfNotNull(
+    result,
+    'supportedCpuArchitectures',
+    Device.supportedCpuArchitectures,
+  );
+  setIfNotNull(result, 'totalMemory', Device.totalMemory);
+
+  const isEmulator = !Device.isDevice;
+  if (isEmulator) result.isEmulator = true;
+
+  return result;
+}
+
+const isAppActive = (appState: AppStateStatus): boolean =>
```

Comment:
Complete personal preference: just write `appState === 'active'` in the code. Adding this helper function means that a reviewer / maintainer needs to look up what this helper does, and it doesn't really save many characters.

Reply:
Removed in f94bdc53aca2fee09b3144baffd35805233f051b.